### PR TITLE
Kb TargetedMetrics Coverage Ceiling

### PR DESCRIPTION
**Pull request by kbergin**
The previous version used shorts instead of ints and we were running into issues of more than MAX_SHORT reads hitting a given target. Changed to ints and included a test to show that more than MAX_SHORT reads can now hit a target with no issue.

@meganshand has recently hit an array index out of bounds bug in this same code and we are working to figure that out still, but I've left a test in this pull request that is ready to modify to test for the bug she is working on as there were no tests for this code before.

### CODE REVIEW

**Comment by kbergin** [comment]
This is in response to ticket #232

File: src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        header.addReadGroup(readGroupRecord);
+        setBuilder.setReadGroup(readGroupRecord);
+        setBuilder.setUseNmFlag(true);
+
+        setBuilder.setHeader(header);
+
+        for (int i = 0; i < numReads; i++) {
+            final int start = rg.nextInt(max) + min;
+            final String newReadName = readName + separator + ID + separator + i;
+            setBuilder.addPair(newReadName, 0, start+ID, start+ID+99);
```
Comment (yfarjoun) [inline src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java]: extract 99 as a constant

File: src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                "AMPLICON_INTERVALS=" + targetIntervals
+        };
+
+        Assert.assertEquals(runPicardCommandLine(args), 0);
+
+        final MetricsFile<TargetedPcrMetrics, Comparable<?>> output = new MetricsFile<TargetedPcrMetrics, Comparable<?>>();
+        output.read(new FileReader(outfile));
+
+        for (final TargetedPcrMetrics metrics : output.getMetrics()) {
+            Assert.assertEquals(metrics.TOTAL_READS, 80000);
```
Comment (yfarjoun) [inline src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java]: numReads*2

File: src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        runMaxReadsTest(tempSamFile);
+        runEmptyIntervalTest(tempSamFile);
+    }
+
+    public void runMaxReadsTest(final File input) throws IOException {
+        final File outfile = File.createTempFile("test", ".TargetedMetrics_Coverage");
+        final File perTargetOutfile = File.createTempFile("perTarget", ".perTargetCoverage");
+        final String referenceFile = "testdata/picard/quality/chrM.reference.fasta";
+        final String targetIntervals = "testdata/picard/quality/chrM.single.interval_list";
+
```
Comment (yfarjoun) [inline src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java]: perTargetOutfile.deleteOnExit()

File: src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    public void runEmptyIntervalTest(final File input) throws IOException {
+        final File outfile = File.createTempFile("test", ".TargetedMetrics_Coverage");
+        final File perTargetOutfile = File.createTempFile("perTarget", ".perTargetCoverage");
+        final String referenceFile = "testdata/picard/quality/chrM.reference.fasta";
+        final String targetIntervals = "testdata/picard/quality/chrM.empty.interval_list";
+
+        outfile.deleteOnExit();
+
```
Comment (yfarjoun) [inline src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java]: perTargetOutfile.deleteOnExit()

File: testdata/picard/quality/chrM.empty.interval_list [hunk 5]
```diff
@@ -0,0 +1,4 @@
+@HD	VN:1.0	SO:coordinate
+@SQ	SN:chrM	LN:16903	AS:mm9	UR:ftp://hgdownload.cse.ucsc.edu/goldenPath/Mus_musculus_assembly9/bigZips/chromFa.tar.gz	M5:11c8af2a2528b25f2c080ab7da42edda	SP:Mus	musculus
```
Comment (yfarjoun) [inline testdata/picard/quality/chrM.empty.interval_list]: Given that the reference we actually use is not the one listed in the ftp site, could you change this? (below as well)

File: src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        BufferedLineReader bufferedLineReader = null;
+        try {
+            bufferedLineReader = new BufferedLineReader(new FileInputStream(dict));
+        } catch (final FileNotFoundException e) {
+            e.printStackTrace();
+        }
+
+        final SAMTextHeaderCodec codec = new SAMTextHeaderCodec();
+        final SAMFileHeader header = codec.decode(bufferedLineReader, dict.toString());
```
Comment (yfarjoun) [inline src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java]: Seems like a very complicated way to initialize a samFileHeader...why not the following:

    header.setSequenceDictionary(SAMSequenceDictionaryExtractor.extractDictionary(dict))

?

File: src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    @Test //test all collection levels
+    public void test() throws IOException{
+        runMaxReadsTest(tempSamFile);
+        runEmptyIntervalTest(tempSamFile);
+    }
+
+    public void runMaxReadsTest(final File input) throws IOException {
+        final File outfile = File.createTempFile("test", ".TargetedMetrics_Coverage");
+        final File perTargetOutfile = File.createTempFile("perTarget", ".perTargetCoverage");
+        final String referenceFile = "testdata/picard/quality/chrM.reference.fasta";
```
Comment (yfarjoun) [inline src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java]: Better extract this to the top of the class, you are also using this in a different functions below.

File: src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+                .setCreateIndex(true).makeBAMWriter(header, false, tempSamFileUnsorted);
+
+        for (final SAMRecord record : setBuilder) {
+            writer.addAlignment(record);
+        }
+        writer.close();
+
+        //sort the temp file
+        final SortSam sorter = new SortSam();
+        final String[] args = new String[]{"INPUT=" + tempSamFileUnsorted.getAbsolutePath(), "OUTPUT=" + tempSamFile.getAbsolutePath(), "SORT_ORDER=coordinate"};
```
Comment (yfarjoun) [inline src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java]: align as you did below, to make easier to read.

File: src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+        sorter.instanceMain(args);
+    }
+
+    @Test //test all collection levels
```
Comment (yfarjoun) [inline src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java]: Better make each of the 2 functions a @Test and use a @DataProvider providing {tempSamFile} to both of them. (easier to see which test fails if it does)

File: src/java/picard/analysis/directed/TargetMetricsCollector.java [hunk 10]
```diff
@@ -589,19 +589,19 @@ public void addMetricsToFile(MetricsFile<METRIC_TYPE, Integer> hsMetricsComparab
...
         public Coverage(final Interval i, final int padding) {
             this.interval = i;
-            this.depths = new short[interval.length() + 2*padding];
+            this.depths = new int[interval.length() + 2*padding];
         }
 
         /** Adds a single point of depth at the desired offset into the coverage array. */
         public void addBase(final int offset) {
             if (offset >= 0 && offset < this.depths.length) {
                 // Prevent overflow if depth is too great, while avoiding doubling memory requirement.
```
Comment (yfarjoun) [inline src/java/picard/analysis/directed/TargetMetricsCollector.java]: comment is false now.

File: testdata/picard/quality/chrM.empty.interval_list (hunk 5 above)
Comment (yfarjoun) [inline testdata/picard/quality/chrM.empty.interval_list]: This is not an empty interval. To get an empty interval use an end position that equals start-1:

     chrM    10    9   +    empty-interval!

(but the test will now fail...)

**Comment by yfarjoun** [comment]
review done. thanks! back to you.

**Comment by kbergin** [comment]
@yfarjoun responded to your comments. I found another bug with interval lists with one interval of length zero (which I don't know why this would ever be the scenario), so I put in an if statement and told it to throw a picard exception if this ever did come up.

The empty interval list test is mostly testing that the program runs without failing. It outputs a warning to the log file but I couldn't figure out how to test that that warning gets output, is there a way to do that?

File: src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        final String separator = ":";
+        final int ID = 1;
+        final int max = 15000;
+        final int min = 1;
+        final Random rg = new Random(5);
+
+        for (int i = 0; i < numReads; i++) {
+            final int start = rg.nextInt(max) + min;
+            final String newReadName = readName + separator + ID + separator + i;
+            final int LENGTH = 99;
```
Comment (yfarjoun) [inline src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java]: extract Length from loop. capitalization is fitting for a static variable, so put it all the way up in the class.

File: src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        final SortSam sorter = new SortSam();
+        final String[] args = new String[]{
+                "INPUT=" + tempSamFileUnsorted.getAbsolutePath(),
+                "OUTPUT=" + tempSamFile.getAbsolutePath(),
+                "SORT_ORDER=coordinate"
+        };
+
+        sorter.instanceMain(args);
+
+
```
Comment (yfarjoun) [inline src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java]: spurious empty line

File: src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        //Write SAM file
+        final SAMFileWriter writer = new SAMFileWriterFactory()
+                .setCreateIndex(true).makeBAMWriter(header, false, tempSamFileUnsorted);
+
+        for (final SAMRecord record : setBuilder) {
+            writer.addAlignment(record);
+        }
+        writer.close();
+
```
Comment (yfarjoun) [inline src/tests/java/picard/analysis/directed/CollectTargetedMetricsTest.java]: spurious empty line

File: src/java/picard/util/MathUtil.java [hunk 14]
```diff
@@ -144,6 +144,16 @@ public static double min(final double[] nums) {
...
+    /** Returns the smallest value stored in the array. */
+    public static int min(final int[] nums) {
+        int min = nums[0];
+        for (int i = 1; i < nums.length; ++i) {
+            if (nums[i] < min) min = nums[i];
+        }
+
+        return min;
+    }
+
```
Comment (yfarjoun) [inline src/java/picard/util/MathUtil.java]: spurous empty line

**Comment by yfarjoun** [comment]
looks good! squash the commits and merge when test pass! :+1:

**Comment by yfarjoun** [comment]
(with my minor comment...)

**Comment by kbergin** [comment]
@meganshand merged!

**Pull request merged**